Providers/Profile.swift

```diff
@@ -436,8 +441,33 @@ open class BrowserProfile: Profile {
     }
 
     lazy var placesDbPath = URL(fileURLWithPath: (try! files.getAndEnsureDirectory()), isDirectory: true).appendingPathComponent("places.db").path
-
-    lazy var places = RustPlaces(databasePath: placesDbPath)
+    lazy var browserDbPath =  URL(fileURLWithPath: (try! self.files.getAndEnsureDirectory())).appendingPathComponent("browser.db").path
+    lazy var places: RustPlaces = RustPlaces(databasePath: self.placesDbPath)
+
+    public func migrateHistoryToPlaces(migrationConfig: HistoryMigrationConfiguration, callback: @escaping (HistoryMigrationResult) -> Void, errCallback: @escaping (Error?) -> Void) {
+        let lastSyncTimestamp = Int64(self.syncManager.lastSyncFinishTime ?? 0)
+        switch migrationConfig {
+        case .disabled:
+            break
+        case .dryRun:
+            let dryRunPath = URL(fileURLWithPath: (try! self.files.getAndEnsureDirectory())).appendingPathComponent("dry-run-places.db").path
+            RustPlaces(databasePath: dryRunPath)
+                .migrateHistory(
+                    dbPath: self.browserDbPath,
+                    lastSyncTimestamp: lastSyncTimestamp,
+                    completion: { result in
+                    do {
+                        try FileManager.default.removeItem(atPath: dryRunPath)
```

Comment:
If the migration never completes, isn't there a risk that this file will exist when you next try? Maybe migration should remove the target DB at the start too? (Or maybe the migration code in a-s actually handles this case OK?)

Reply:
Yup, if the migration errors out or doesn't complete then the file will still be there - but we do handle it properly (the db will be empty next time since everything is in one transaction!) and **eventually** we'll delete the database once the migration passes